sql: introduce null_ordered_last session variable

otan opened the pull request:
See individual commits for details.
Resolves #71381

otan committed:
sql: introduce null_ordered_last session variable

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71429)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
LGTM, except for one thing: in Memo (opt/memo/memo.go), we keep track of all settings that can affect planning. This is necessary to avoid reusing cached memos when the settings change. Grep for e.g. `intervalStyleEnabled` in `memo.go` and `memo_test.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71429#-:-MlmWQwZ3UmRbphCbvEE:b7jxng4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

otan committed:
optbuilder: use null_ordered_last session variable

Release note (sql change): NULLS can be ordered NULLS LAST by default
if the `null_ordered_last` session variable be set to true.

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71429#-:-MlqR-UXBzYTHJ_izgXD:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/opt/memo/memo.go, line 136 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71429#-MlqQeRT-44MGR933s4S:-MlqQeRT-44MGR933s4T:b-oz0a1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7912d665e792ac8bb9d71608b93f784f56b3a95/pkg/sql/opt/memo/memo.go#L136)):*
> ```Go
> 	// The following are selected fields from SessionData which can affect
> 	// planning. We need to cross-check these before reusing a cached memo.
> 	// NOTE: If you add new fields here, be sure to add them to the relevant
> ```

Do this too please (line 527 or so of explain_bundle.go)


<!-- Sent from Reviewable.io -->

otan commented:
bors r=RaduBerinde
